File: Server/GenerateTableClass.py
import os
import codecs

out1 = """
/** 包含的头文件 */
%s

/**
 * [当前为生成代码，不可以修改] CSC 配置表的管理类
 * */
var CSVManagerInstance = (function() {
    /** 类的成员变量 */
    var _unique;
%s
    /**
    * 单例函数
    */
    function Instance() {
        if(_unique === undefined) {
            _unique = new CSVManager();
        }
        return _unique;
    }

    /**
    * 构造函数
    */
    function CSVManager() {
%s
    }
%s
    /**
    * 返回一个单例函数
    */
    return Instance;
})();

/** 声明一个单例对象 */
exports.Instance = CSVManagerInstance;
"""

out2 = """
/** 包含的头文件 */
var csvParse = require('../tools/parse/csv');

/**
 * [当前为生成代码，不可以修改] {0} 类为 {0}Table 每一行的元素对象
 * */
var {0} = (function() {{

    /**
    * 构造函数
    */
    function {0}() {{
{1}
    }}

    return {0};
}})();

/**
 * [当前为生成代码，不可以修改] {0} 配置表
 * */
var {0}TableInstance = (function() {{

    /**
    * 类的成员变量
    */
    var _unique;
    var _lines;

    /**
    * 单例函数
    */
    function Instance() {{
        if(_unique === undefined) {{
            _unique = new {0}Table();
        }}
        return _unique;
    }}

    /**
    * 构造函数
    */
    function {0}Table() {{
        _lines = {{}};
        var tmpArr = csvParse.Instance().getTable('{0}');
        for(var i = 0; i < tmpArr.length; ++i) {{
            var obj = new {0}();
{2}
            _lines[{4}] = obj;
{3}
        }}
    }}

    /**
    * 获取行队列
    */
    {0}Table.prototype.GetLines = function() {{
        return _lines;
    }};
    
    /**
    * 返回一个单例函数
    */
    return Instance;
}})();

/**
* 声明一个单例对象
*/
exports.Instance = {0}TableInstance;
"""

dirName = "data";
rroot = os.path.join(os.getcwd(), 'csvTables')
a1 = []
a2 = []
a3 = []
a4 = []

table_js_root = os.path.join(os.getcwd(), "server", "data")
if not os.path.isdir(table_js_root): exit(0)

manager_js_root = os.path.join(os.getcwd(), "server", "manager")
if not os.path.isdir(manager_js_root): exit(0)

h = open(os.path.join(manager_js_root, 'csv_manager.js'), 'w', encoding = "utf8")


for root, subdirs, files in os.walk(rroot):
    for file in files:
        if(file.find('.csv') == -1): continue
        filename = file.split(".")[0]

        a1.append("var csv" + filename.upper() + " = require('../"+dirName+"/" + filename + "');\n");
        a2.append("    var _" + filename + ";\n");
        a3.append("        _" + filename +  " = csv" + filename.upper() + ".Instance;\n");
        a4.append("    CSVManager.prototype." + filename + " = function() {\n" + "        return _" + filename + "().GetLines();\n    };\n\n")

        f = codecs.open(os.path.join(root, file), 'r', encoding = "utf_8_sig")
        g = open(os.path.join(table_js_root, filename + '.js'), 'w', encoding = "utf8")

        line = f.readline().strip()
        nm = line.split(",")

        line = f.readline().strip()
        tp = line.split(",")

        ln = len(nm)

        b1 = []
        b2 = []
        b3 = []
        field = '';
        for i in range(ln):
            if i == 0:
                field = 'tmpArr[i].' + nm[i];
            if tp[i] == 'INT':
                b1.append("        this." + nm[i] + " = 0;\n")
                b2.append("            obj." + nm[i] + " = parseInt(tmpArr[i]." + nm[i] + ");\n")
                
            elif tp[i] == "FLOAT":
                b1.append("        this." + nm[i] + " = 0;\n")
                b2.append("            obj." + nm[i] + " = parseFloat(tmpArr[i]." + nm[i] + ");\n")

            elif tp[i] == "STRING":
                b1.append("        this." + nm[i] + " = '';\n")
                b2.append("            obj." + nm[i] + " = tmpArr[i]." + nm[i] + ";\n")
                
        b1 = "".join(b1)
        b2 = "".join(b2)
        b3 = "".join(b3)

        
        g.write(out2.format(filename, b1, b2, b3, field))
        f.close()
        g.close()


a1 = "".join(a1)
a2 = "".join(a2)
a3 = "".join(a3)
a4 = "".join(a4)
h.write(out1 % (a1, a2, a3, a4))
h.close()

File: Server/generateJS/GenerateTableClass.js
/**
 * Created by 薄雪婷 on 2016/3/8.
 */
var fs = require('fs');
var path = require('path');

var root = path.join(process.cwd(), 'csvTables');                               //当前的绝对路径
var table_js_root = path.join(process.cwd(), "server", "data");                   //要读取的文件夹的路径
var manager_js_root = path.join(process.cwd(), "server", "manager");

var out1 = "\n\
/** 包含的头文件 */\n\
{0}\n\
\n\
/**\n\
 * [当前为生成代码，不可以修改] CSC 配置表的管理类\n\
 * */\n\
var CSVManagerInstance = (function() {\n\
    /** 类的成员变量 */\n\
    var _unique;\n\
{1}\n\
    /**\n\
     * 单例函数\n\
     */\n\
    function Instance() {\n\
        if(_unique === undefined) {\n\
            _unique = new CSVManager();\n\
        }\n\
        return _unique;\n\
    }\n\
\n\
    /**\n\
     * 构造函数\n\
     */\n\
    function CSVManager() {\n\
{2}\n\
    }\n\
{3}\n\
    /**\n\
     * 返回一个单例函数\n\
     */\n\
    return Instance;\n\
})();\n\
\n\
/** 声明一个单例对象 */\n\
exports.Instance = CSVManagerInstance;\n\
";

var out2 = "\n\
/** 包含的头文件 */\n\
var csvParse = require('../tools/parse/csv');\n\
\n\
/**\n\
 * [当前为生成代码，不可以修改] {0} 类为 {0}Table 每一行的元素对象\n\
 * */\n\
var {0} = (function() {\n\
\n\
    /**\n\
     * 构造函数\n\
     */\n\
    function {0}() {\n\
{1}\n\
    }\n\
\n\
    return {0};\n\
})();\n\
\n\
/**\n\
 * [当前为生成代码，不可以修改] {0} 配置表\n\
 * */\n\
var {0}TableInstance = (function() {\n\
\n\
    /**\n\
     * 类的成员变量\n\
     */\n\
    var _unique;\n\
    var _lines;\n\
\n\
    /**\n\
     * 单例函数\n\
     */\n\
    function Instance() {\n\
        if(_unique === undefined) {\n\
            _unique = new {0}Table();\n\
        }\n\
        return _unique;\n\
    }\n\
\n\
    /**\n\
     * 构造函数\n\
     */\n\
    function {0}Table() {\n\
        _lines = {};\n\
        var tmpArr = csvParse.Instance().getTable('{0}');\n\
        for(var i = 0; i < tmpArr.length; ++i) {\n\
            var obj = new {0}();\n\
{2}\n\
            _lines[{4}] = obj;\n\
{3}\n\
        }\n\
    }\n\
\n\
    /**\n\
     * 获取行队列\n\
     */\n\
    {0}Table.prototype.GetLines = function() {\n\
        return _lines;\n\
    };\n\
\n\
    /**\n\
     * 返回一个单例函数\n\
     */\n\
    return Instance;\n\
})();\n\
\n\
/**\n\
 * 声明一个单例对象\n\
 */\n\
exports.Instance = {0}TableInstance;\n\
";

/**
 * 格式化字符串
 */
String.prototype.format = function (args) {
    var ret = this;
    if(arguments.length > 0) {
        for(var i = 0; i < arguments.length; ++i) {
            var reg = new RegExp('({)' + i + '(})', 'g');
            ret = ret.replace(reg, arguments[i]);
        }
    }
    return ret;
};

/**
 * 读取文件，将文件内容放入数组m中,参数pathname即为文件的路径
 */
function readFile(pathname, callback) {
    //var m = [];
    var input = fs.readFileSync(pathname, 'utf-8').trim().split("\n");
    //m = (input.split("\r\n")).slice(0,(input.split("\r\n")).length - 1);
    callback(input);
}

/**
 * 创建js文件，并写入数据
 */
function createJSFile(pathname) {
    var a1 = [];
    var a2 = [];
    var a3 = [];
    var a4 = [];
    fs.readdirSync(pathname).forEach(function (fileName) {
        var fullPath = path.join(pathname, fileName);/* 拼接完整路径 */
        var filename = '';
        if (fs.statSync(fullPath).isFile() && (fileName.indexOf('.csv') !== -1)) {
            filename = path.basename(fullPath, '.csv');                         //filename是不加后缀名的文件名称
        }

        if (filename !== '') {
            a1.push("var csv" + filename.toUpperCase() + " = require('../" + "data" + "/" + filename + "');\n");
            a2.push("    var _" + filename + ";\n");
            a3.push("        _" + filename + " = csv" + filename.toUpperCase() + ".Instance;\n");
            a4.push("    CSVManager.prototype." + filename + " = function() {\n" + "        return _" + filename + "().GetLines();\n    };\n\n");

            var b1 = [];
            var b2 = [];
            var b3 = [];
            var field = '';

            readFile(fullPath, function (fileContent) {
                var row = [];                                   //第一行的信息
                var row1 = [];                                  //第二行的信息
                row = fileContent[0].split(",");
                row1 = fileContent[1].split(",");

                for (var j = 0; j < row.length; j++) {
                    if (j === 0) {
                        field = 'tmpArr[i].' + row[j];
                    }

                    if (row1[j] === 'INT') {
                        b1.push("        this." + row[j] + " = 0;\n");
                        b2.push("            obj." + row[j] + " = parseInt(tmpArr[i]." + row[j] + ");\n");
                    } else if (row1[j] === "FLOAT") {
                        b1.push("        this." + row[j] + " = 0;\n");
                        b2.push("            obj." + row[j] + " = parseFloat(tmpArr[i]." + row[j] + ");\n");
                    } else if (row1[j] === "STRING") {
                        b1.push("        this." + row[j] + " = '';\n");
                        b2.push("            obj." + row[j] + " = tmpArr[i]." + row[j] + ";\n");
                    }
                }

            });
            var b11 = b1.join("");
            var b22 = b2.join("");
            var b33 = b3.join("");

            fs.writeFile(path.join(table_js_root, filename + '.js'), (out2.format(filename, b11, b22, b33, field)), {encoding: 'utf-8'}, function (err) {
                if (err) {
                    console.error(err);
                } else {
                    console.log("写入 " + filename + ".js 成功");
                }
            });
        }
    });
    var a11 = a1.join("");
    var a22 = a2.join("");
    var a33 = a3.join("");
    var a44 = a4.join("");
    fs.writeFile(path.join(manager_js_root, 'csv_manager.js'), (out1.format(a11, a22, a33, a44)), { encoding:'utf-8' }, function (err) {
        if(err){
            console.error(err);
        }else{
            console.log("写入 csv_manager.js 成功");
        }
    });
}

function startWrite(pn){
    if((!(fs.statSync(table_js_root).isDirectory())) || (!(fs.statSync(manager_js_root).isDirectory()))){
        console.log("要写入的目标文件夹不存在!!!");
        return;
    }
    createJSFile(pn);
}

startWrite(root);
